test/runnable/test14650.d

```diff
@@ -0,0 +1,16 @@
+int a = 0;
+
+struct S
+{ ~this() { ++a; } }
+
+
+static ~this()
+{
+    assert(a == 2);
```

Comment:
I think the globals should be destroyed after the explicit `static ~this()` have run, so the test should verify that (a == 0).
A `shared static ~this()` can verify the destruction instead.

Reply:
Is it possible to tell if a static destructor accesses a particular variable? That probably would be a more complex fix.

Reply:
> This reveals a dilemma. a `shared static ~this()` is going to currently expect it's own thread local storage to still be valid. Changing this will break that code.

Maybe the main thread should run TLS destructors after `shared static ~this()`, but before shared destructors. The latter could contain the assertion above.

Reply:
Isn't `shared static ~this()` a shared destructor?

Reply:
> Is it possible to tell if a static destructor accesses a particular variable? That probably would be a more complex fix.

It is possible to know if a particular variable is accessed, but it is hard to tell if the variable is destroyed or not:

```d
struct A
{
     ~this() {}
}

A a;

void fun(ref A a)
{
     a.__dtor();
}

static ~this()
{
    import std.stdio : writeln;
    writeln(a);    // accessed but not destroyed;
}

static ~this()
{
    fun(a);       // need to examine fun to tell if it is destroyed or not
}
```

Not to mention the overhead of re-analyzing all the function bodies that a static destructor calls.

Honestly, at this point I think that a better approach would be to document the existing behavior and guide the user to destroy its global variables using a static destructor. I'm not really sure how you can destroy static variables from different scopes (maybe use the mangled name?).

Reply:
>>but before shared destructors. 
> Isn't `shared static ~this()` a shared destructor?

I meant destructors of shared variables.

>You mean that shared variables should be destroyed in a shared static destructor?

yes, if it is the last to be executed for that module.

Reply:
> Honestly, at this point I think that a better approach would be to document the existing behavior and guide the user to destroy its global variables using a static destructor.

There are some issues there:

1) adding a static destructor to a module makes it participate in the cycle detection. It may not be possible to add such things. However, we can in the compiler make standalone static destructors which are free to execute in any order (that change needs to go into this PR).
2) any destructors should be run AFTER everything is done, doing it in the compiler the way this PR is done is much safer than someone adding a static destructor, possibly in the middle of the module, that will destroy the item early.

Reply:
> Maybe the main thread should run TLS destructors after `shared static ~this()`, but before shared destructors. The latter could contain the assertion above.

This means adding another entry point to moduleinfo, as all destructors are combined into one function.